src/Features/Core/Portable/NavigateTo/NavigateToItem.cs

```diff
@@ -0,0 +1,272 @@
+
+        [DataMember(Order = 8)]
+        public readonly TextSpan DeclaredSymbolInfoSourceSpan;
+
+        [DataMember(Order = 9)]
+        public readonly bool DeclaredSymbolInfoIsPartial;
+
+        [DataMember(Order = 10)]
+        public readonly bool DeclaredSymbolInfoIsNonNestedNamedType;
+
+        [DataMember(Order = 11)]
+        public readonly string Kind;
+
+        [DataMember(Order = 12)]
+        public readonly NavigateToMatchKind MatchKind;
+
+        [DataMember(Order = 13)]
+        public readonly bool IsCaseSensitive;
+
+        [DataMember(Order = 14)]
+        public readonly ImmutableArray<TextSpan> NameMatchSpans;
+
+        [DataMember(Order = 15)]
+        public readonly string SecondarySort;
+
+        public RoslynNavigateToItem(
+            bool isStale,
+            DocumentId documentId,
+            string combinedProjectName,
+            string declaredSymbolInfoName,
+            string declaredSymbolInfoNameSuffix,
+            string declaredSymbolInfoContainerDisplayName,
+            DeclaredSymbolInfoKind declaredSymbolInfoKind,
+            Accessibility declaredSymbolInfoAccessibility,
+            TextSpan declaredSymbolInfoSourceSpan,
+            bool declaredSymbolInfoIsPartial,
+            bool declaredSymbolInfoIsNonNestedNamedType,
+            string kind,
+            NavigateToMatchKind matchKind,
+            bool isCaseSensitive,
+            ImmutableArray<TextSpan> nameMatchSpans,
+            string secondarySort)
+        {
+            IsStale = isStale;
+            DocumentId = documentId;
+            CombinedProjectName = combinedProjectName;
+            DeclaredSymbolInfoName = declaredSymbolInfoName;
+            DeclaredSymbolInfoNameSuffix = declaredSymbolInfoNameSuffix;
+            DeclaredSymbolInfoContainerDisplayName = declaredSymbolInfoContainerDisplayName;
+            DeclaredSymbolInfoKind = declaredSymbolInfoKind;
+            DeclaredSymbolInfoAccessibility = declaredSymbolInfoAccessibility;
+            DeclaredSymbolInfoSourceSpan = declaredSymbolInfoSourceSpan;
+            DeclaredSymbolInfoIsPartial = declaredSymbolInfoIsPartial;
+            DeclaredSymbolInfoIsNonNestedNamedType = declaredSymbolInfoIsNonNestedNamedType;
+            Kind = kind;
+            MatchKind = matchKind;
+            IsCaseSensitive = isCaseSensitive;
+            NameMatchSpans = nameMatchSpans;
+            SecondarySort = secondarySort;
+        }
+
+        public bool TryCreateSearchResult(Solution solution, [NotNullWhen(true)] out INavigateToSearchResult? result)
+        {
+            var document = solution.GetDocument(DocumentId);
+            if (document == null)
+            {
+                var project = solution.GetProject(DocumentId.ProjectId);
+                if (project != null)
+                    document = project.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(DocumentId);
+            }
```

Comment:
@jasonmalinowski is this the right way to map a doc id back to a possibly source-generated document?

Reply:
Is it possible to just make this path async? Because then you can just call GetDocumentAsync().

Reply:
And the longer answer: that API I _really_ wish I could mark as friend of the other parts of the workspace so it wasn't available but we don't have that in C#. This might work though, but asyncifying is preferrable.

Reply:
> Is it possible to just make this path async? Because then you can just call GetDocumentAsync().

i don't want this to be async.  if it references a source generated document, then we better have that document.  i do *not* want the documents to be generated.